### ISSUE TITLE
Separate dev and production dependencies

### ISSUE THREAD

**Issue by urbanautomaton** [Lostmyname/lmn-gulp-tasks#1]
A problem we're seeing now that we've switched to lmn-gulp-tasks in mermaid is that all dependencies are at the production level, so we're back to installing e.g. jscs, jshint, browser-sync etc. on heroku.

Is it possible to move dependencies of development-only gulp tasks into a devDependencies group?

**Comment by callumacrae** [comment]
You can't just load some gulp tasks and not others. That's gulp, not lmn-gulp-tasks, and it's never been any different.

You can load lmn-gulp-tasks as a devDependency, but then you've only got it locally.

**Comment by urbanautomaton** [comment]
Yep, understood; but it seems like it's possible to defer when the gulp tasks load their dependencies until they're actually invoked. So for example in `js-quality.js`, where currently you've got this:

``` js
'use strict';

var stylish = require('jshint-stylish');

module.exports = function (gulp, plugins, options) {
  return function jsQualityTask() {
    var stream = gulp.src(options.src);
    // the actual task
};
```

If you had this:

``` js
'use strict';

module.exports = function (gulp, plugins, options) {
  return function jsQualityTask() {
    var stream = gulp.src(options.src);
    var stylish = require('jshint-stylish');
    // the actual task
};
```

This means you can load the js-quality task without it erroring, but will receive an error if you attempt to invoke it without jshint-stylish being available.

**Comment by urbanautomaton** [comment]
Ahem, I left an errant `require` call in the second example.

**Comment by callumacrae** [comment]
Putting `require` anywhere other than the top of the file is pretty bad practice.

---

Also, that means we're deciding in _this_ repository which tasks should be ran on development and which ones on production, and one day if we decide that we want to run js-quality on production, we would have to change this repository, which would change it on every other repository it's used in.

It feels like a hack, and it's not nice at all.

**Comment by callumacrae** [comment]
Also, it would break all the components. devDependencies aren't "for development environments", they're for developing modules. They're not installed when lmn-gulp-tasks is a dependency of another module—such as a component.

**Comment by urbanautomaton** [comment]
Without disagreeing with any of what you've said, we do still have a strong incentive not to install every dependency of every gulp task in this repo on heroku, so I'd like to find a solution if at all possible.

So, for example, presently `js-quality` is very much a development dependency of the mermaid module. If it's bundled with all the other tasks, then there's certainly no way to load it for development of mermaid without also installing it in production. Would splitting the tasks into separate modules make sense, for example, so I could specify a full dependency on some tasks, and a dev dependency on others?

**Comment by callumacrae** [comment]
Feel free to disagree with what I've said :smile: 

Splitting it into separate modules would work, but with ten tasks, a loader file, and a rev helper, that's twelve modules, all of which would be separate repositories. That's even less maintainable than just keeping the gulp tasks in every repository! Adding sourcemaps, which we probably want to do at some point, would be difficult and would result in duplicate dependencies—half the problem this project was designed to solve.

When we come to add tests to `lmn-gulp-tasks`, right now it's pretty easy—I just create a test directory and write some tests. If everything were in separate modules, there would need to a repository dedicated to testing the tasks—and I'm not even sure how that would work when making changes to tasks.

Basically, it's possible, but it's hell.
